src/analysis-cache.ts

```diff
@@ -0,0 +1,169 @@
+/**
+ * @license
+ * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Document, Import, InlineDocument, ScannedDocument, ScannedImport, ScannedInlineDocument} from './model/model';
+import {ParsedDocument} from './parser/document';
+
+export class AnalysisCache {
+  parsedDocumentPromises = new Map<string, Promise<ParsedDocument<any, any>>>();
+  scannedDocumentPromises = new Map<string, Promise<ScannedDocument>>();
+  analyzedDocumentPromises = new Map<string, Promise<Document>>();
+  dependenciesScanned = new Map<string, Promise<void>>();
+
+  scannedDocuments = new Map<string, ScannedDocument>();
+  analyzedDocuments = new Map<string, Document>();
+
+  /**
+   * Returns a copy of this cache, with the given path and all of its transitive
+   * dependants invalidated.
+   */
+  onPathChanged(path: string, dependants: Iterable<string>): AnalysisCache {
```

Comment:
How much complexity does this method add to the rest of the cache/analyzer? My bet is that paths are changed very rarely, and it could reduce a ton of complexity around tracking these different things if we just cleared the cache and rescanned whenever a path is changed.